### PR TITLE
Feature: Added useDirectusAuthHook for State-Based Authentication

### PR DESCRIPTION
**Pull request by Fritsch-Tech**
#### 🔧 Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Codebase improvement
- [ ] Other (if none of the other choices apply)

#### 🚨 Checklist

- [x] I've read the [guidelines for contributing](https://github.com/gremo/react-directus/blob/main/CONTRIBUTING.md)
- [x] I've added necessary documentation (if appropriate)
- [x] I've ensured that my code additions do not fail linting or unit tests (if applicable)

#### Description

This pull request introduces a new feature, `useDirectusAuthHook`, which enables React state-based access to the current user and authentication state. The hook provides convenient methods for direct login and logout, utilizing the context-based states.

The following changes have been made:

1. **useDirectusAuthHook**: The new hook has been implemented to manage user authentication state and user information, allowing easier and more flexible access to this data within the application.

2. **Direct Login and Logout Methods**: The hook includes direct login and logout methods that work seamlessly with the context-based states, simplifying the authentication process.

This feature enhances the overall user experience and provides developers with a more intuitive way to handle authentication within the application. Furthermore, it will seamlessly integrate with the authentication system of the components once it's introduced.

Please review the changes and provide feedback.

Thank you for considering this pull request, and I look forward to your valuable feedback!

### CODE REVIEW

File: src/hooks/useDirectusAuth.tsx [hunk 1]
```diff
@@ -0,0 +1,94 @@
+import { DirectusAuthHook } from '../types';
+import { DirectusContext } from '../DirectusProvider';
+import React from 'react';
+import { UserType } from '@directus/sdk';
+
+/**
+ * A hook to access the Directus authentication state and methods.
+ * @example
```
Comment (gremo) [inline src/hooks/useDirectusAuth.tsx:9]: Maybe this example lacks indentation?

File: src/DirectusProvider.tsx [hunk 2]
```diff
@@ -22,12 +24,18 @@ export const DirectusContext = React.createContext<DirectusContextTypeGeneric<an
...
   children,
 }: DirectusProviderProps): JSX.Element => {
+  const [user, setUser] = React.useState<UserType | null>(null);
+  const [authState, setAuthState] = React.useState<AuthStates>('loading');
+
+  const directus = React.useMemo(() => new Directus<T>(apiUrl, options), [apiUrl, options]);
+
   const value = React.useMemo<DirectusContextType<T>>(
     () => ({
       apiUrl: apiUrl,
```
Comment (gremo) [inline src/DirectusProvider.tsx]: I think there is no need to repeat apiUrl and directus here

File: src/types.ts [hunk 3]
```diff
@@ -61,9 +61,16 @@ export interface DirectusProviderProps {
...
   options?: DirectusOptions;
+  /**
+   * If `true`, the provider will try to login the user automatically on mount.
+   * @default false
+   */
+  autoLogin?: boolean;
   children: React.ReactNode;
 }
 
+export type AuthStates = 'loading' | 'authenticated' | 'unauthenticated';
```
Comment (gremo) [inline src/types.ts:72]: Why not using an enum here?

File: src/hooks/useDirectusAuth.tsx [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        password,
+      });
+
+      const dUser = (await directus.users.me.read({
+        fields: ['*'],
+      })) as UserType;
+
+      if (dUser) {
+        setDirectusUser(dUser);
+        setAuthState('authenticated');
```
Comment (gremo) [inline src/hooks/useDirectusAuth.tsx:66]: Would be better to use an enum not a magic string?

**Comment by gremo** [comment]
Does the README should be updated too to document the usage?

File: src/hooks/useDirectusAuth.tsx (hunk 1 above)
Comment (Fritsch-Tech) [reply src/hooks/useDirectusAuth.tsx:9]: Im what way?

File: src/types.ts (hunk 3 above)
Comment (Fritsch-Tech) [reply src/types.ts:72]: I don't realy have a firm stance in the Enum vs Union debate. If you prefere Enums i can update the code acordingly.
Should i update the DirectusImage fit and format props too?

File: src/DirectusProvider.tsx (hunk 2 above)
Comment (Fritsch-Tech) [reply src/DirectusProvider.tsx]: yes, you are correct.

**Comment by Fritsch-Tech** [comment]
I thought about updating the README, but I wanted your input in regards to #360 first. In case we adopt the automatic documentation generation the TSDoc string and add a link to the doc file in the README it would be sufficient.

File: src/types.ts (hunk 3 above)
Comment (gremo) [reply src/types.ts:72]: No problem then, let's go for unions.

Comment (Fritsch-Tech) [reply src/types.ts:72]: Okay.
I read a little more into it last night. Enums would sound kinda nice, but if we would use them, we should use them everywhere. Sadly that would be a breaking change.

Maybe for the Directus SDK v11 compatible Version?

**Comment by gremo** [comment]
#360 seems big, I honestly don't know how TSDoc generation works. Would you mind add a few lines in the README, so we can merge this? Then we move to the next PR? 😄

**Comment by Fritsch-Tech** [comment]
No Problem.

**Pull request merged**